pkg/image/admission/imagepolicy/accept.go

```diff
@@ -0,0 +1,111 @@
+package imagepolicy
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/admission"
+	apierrs "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+
+	"github.com/openshift/origin/pkg/api/meta"
+	"github.com/openshift/origin/pkg/image/admission/imagepolicy/rules"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+var errRejectByPolicy = fmt.Errorf("this image is prohibited by policy")
+
+func accept(accepter rules.Accepter, resolver imageResolver, m meta.ImageReferenceMutator, attr admission.Attributes, covered policyAttributes, excludedRules sets.String) error {
+	gr := attr.GetResource().GroupResource()
+	requiresImage := accepter.RequiresImage(gr)
+	resolvesImage := accepter.ResolvesImage(gr)
+
+	errs := m.Mutate(func(image string) (string, error) {
+		ok, newImage, err := acceptImage(accepter, resolver, image, gr, covered, requiresImage, resolvesImage, excludedRules)
+		if err != nil {
+			return image, err
+		}
+		if !ok {
+			return image, errRejectByPolicy
+		}
+		return newImage, nil
+	})
+
+	for i := range errs {
+		if errs[i].Detail == errRejectByPolicy.Error() {
+			errs[i].Type = field.ErrorTypeForbidden
+		} else {
+			errs[i].Detail = fmt.Sprintf("this image is prohibited by policy: %s", errs[i].Detail)
+		}
+	}
+
+	if len(errs) > 0 {
+		glog.V(5).Infof("failed to create: %v", errs)
+		return apierrs.NewInvalid(attr.GetKind().GroupKind(), attr.GetName(), errs)
+	}
+	glog.V(5).Infof("allowed: %#v", attr)
+	return nil
+}
+
+func acceptImage(accepter rules.Accepter, resolver imageResolver, name string, gr unversioned.GroupResource, covered policyAttributes, requiresImage, resolvesImage bool, excludedRules sets.String) (bool, string, error) {
+	// only resolve the image once per pod
+	if covered.Has(name) {
```

Comment:
Since you remember the original name, this could match the original name and skip pinning it to a digest, couldn't it?